### ISSUE TITLE
Stateful Algos

### ISSUE THREAD

**Issue by Delaunay** [Epistimio/orion#358]
# Why

Currently Algos are Stateless and the DB holds all the information to build the state up.
Note that the DB does not actually hold an `algo_state` collection but rather shards of states (i.e trials) that are aggregated together to form the algo state.

Current implementation also rely heavily on the database to forbid duplicates which creates awkward workflow where trials are sampled, sent to the db to make sure they are unique and finally accepted by the algo itself (To accept a trial you need to observe it with a result of None.

This splits the simple task of creating a new trial into many different files with many different failure scenarios, that makes this approach error prone and difficult to debug.

This creates a lot of subtleties that makes it hard for researchers to implement new HPO without a deep understanding of Orion's internal. 

Note that this approach has the advantage that multiple HPOs can be instantiated for the same task on different workers to make them cooperate. Nevertheless, the HPO's role is minor and in fact does not need to be alive most of the time as it is waiting for trials to complete.

# Statefulness

If the algo holds its entire state it can know if a trial is unique without relying on the db.
Creating a trial can be local to the HPO itself. In fact its entire logic can be internal without need for interaction with Orion subsystem.

This makes it easy for the research to create new HPOs and to experiment.

There will be one unique HPO instance per experiment, the full HPO state will be saved inside the DB in case of a critical failure so it can be restarted anywhere else if needed.

Current HPO implementation do not need to be rewritten as they can be turned into a statefull algo using an Adapter like below.

```python
class HPOAdater:
    """Make Orion HPO resume-able"""

    def __init__(self, hpo_name, space, fidelity, *,
                 uids=None, results=None, running=None, instance=0, allow_dup=False, **hpo_config):
        # Force consistency of ordering
        if isinstance(space, (dict, OrderedDict)):
            space = list(space.items())

        space.sort(key=lambda i: i[0])

        self.optimizer_name = hpo_name
        self.optimizer_config = hpo_config

        self.fidelity = fidelity
        self.space = OrderedDict(space)

        # remove fidelity from space
        for k, v in self.space.items():
            if v.startswith('fidelity'):
                self.space.pop(k, None)
                break

        # build space with fidelity for Orion
        self.hpo_space = self._build_hpo_space(self.space, fidelity)

        self.uid_set = set(select(uids, set()))
        self.results = select(results, [])
        self.running = select(running, dict())
        self.allow_duplicate = allow_dup
        self.instance = instance + 1

        self.optimizer = OptimizationAlgorithm(
            hpo_name,
            space=self.hpo_space,
            **hpo_config)

        self._resume()
        self.budget = [
            dict(epoch=epoch, trial_count=trial_count) for trial_count, epoch in self._budget[0]
        ]

    def _build_hpo_space(self, space, fidelity):
        temp_space = copy.deepcopy(space)
        temp_space[fidelity.name] = str(fidelity)
        builder = SpaceBuilder()
        return builder.build(temp_space)

    def _resume(self):
        # Observe old trials
        for uid, params, objective in self.results:
            self._observe(params, objective)

        # re accept currently running trials
        for uid, params in self.running.items():
            self._accept(params)

    @property
    def trial_count(self):
        return sum([data['trial_count'] for data in self.budget])

    @property
    def trial_count_remaining(self):
        return self.trial_count - (len(self.running) + len(self.results))

    def state_dict(self):
        return {
            'space': list(self.space.items()),
            'uids': list(self.uid_set),
            'results': self.results,
            'running': self.running,
            'instance': self.instance,
            'allow_dup': self.allow_duplicate,
            'hpo_name': self.optimizer_name,
            'hpo_config': self.optimizer_config,
            'fidelity': self.fidelity.to_dict()
        }

    def suggest(self) -> List[Tuple['uid', List[Argument]]]:
        """Sample new trials or promote old trials"""
        suggestions = []
        new_suggestions = self._suggest()

        while len(new_suggestions) > 0:
            for uid, params in new_suggestions:
                if self.allow_duplicate or uid not in self.uid_set:
                    self._accept(params)
                    self.uid_set.add(uid)

                    # The trial is already running
                    if uid not in self.running:
                        suggestions.append((uid, params))
                        self.running[uid] = params
                    else:
                        debug(f'refusing {uid} because it is already running')
                else:
                    debug(f'refusing because (allow_dup: {self.allow_duplicate})')
            new_suggestions = self._suggest()

        # first call to suggest should have allow_duplicate == False
        # Subsequent call should have it == True
        if not self.allow_duplicate:
            self.allow_duplicate = True

        return suggestions

    def observe(self, results: List[Tuple['uid', List[Argument], float]]):
        """Observe new trials"""
        for uid, params, objective in results:
            assert self.unique_id(self.params_to_points(params)) == uid, 'uid should be deterministic'

            self._observe(params, objective)
            # cant be pop because in case of twin killer we receive more that one result
            self.running.pop(uid, None)
            self.results.append((uid, params, objective))

    def unique_id(self, points):
        """Compute a unique id from the hyper parameters without fidelity"""
        import hashlib
        sh256 = hashlib.sha256()
        try:
            fidelity_dim = self.optimizer.fidelity_index
        # TwinKiller does not have fidelity_index but we know it is the last
        except AttributeError:
            fidelity_dim = len(points) - 1

        for i, p in enumerate(points):

            if i != fidelity_dim:
                sh256.update(str(p).encode('utf8'))

        return sh256.hexdigest()[0:16]

    # @property
    # def _rung_sizes(self):
    #     reduction_factor = self.fidelity.base
    #     return compute_rung_sizes(reduction_factor, len(self._budget))

    @property
    def _budget(self):
        # min_resources = self.fidelity.low
        max_resources = self.fidelity.max
        reduction_factor = self.fidelity.base

        return compute_budgets(max_resources, reduction_factor)

    def convert(self, val):
        import numpy as np
        if isinstance(val, np.int64):
            return int(val)

        if isinstance(val, np.float):
            return float(val)

        return val

    def points_to_params(self, points):
        params = []

        for name, value in zip(self.space.keys(), points):
            params.append((name, self.convert(value)))

        params.append((self.fidelity.name, points[-1]))
        return params

    def params_to_points(self, params):
        points = []
        params_dict = dict(params)

        for k in self.space.keys():
            points.append(params_dict[k])

        return tuple(points) + (params_dict[self.fidelity.name],)

    def _suggest(self):
        points = self.optimizer.suggest()

        if points is None:
            return []

        return [(self.unique_id(p), self.points_to_params(p)) for p in points]

    def _accept(self, params):
        self.optimizer.observe([self.params_to_points(params)], [{'objective': None}])

    def _observe(self, params, value):
        self.optimizer.observe([self.params_to_points(params)], [{'objective': value}])

    @property
    def is_done(self):
        return self.optimizer.is_done
```

**Comment by bouthilx** [comment]
I totally agree! This is the next step we should take in the refactoring of the algorithm interface. I don't have time to look at the proposed implementation in details now but we should be ready to start working on this in a few weeks.
